### PR TITLE
Improvements in Database metadata to prevent Statement leaks and enhance Statement caching

### PR DESCRIPTION
**Pull request by cheenamalhotra**
This PR introduces below changes to SQLServerDatabaseMetadata:

- Improved Statement caching for different catalogs
- Enables Statement caching for 'null' catalog
- Enables closing un-cached Statements on completion of ResultSet _(Prevents Statement Leaks)_

### CODE REVIEW

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCopy.java [hunk 1]
```diff
@@ -1768,8 +1766,6 @@ private void getDestinationMetadata() throws SQLServerException {
                 rs.close();
             if (null != stmt)
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCopy.java]: why not try-with-resouce for this as well?

Comment (cheenamalhotra) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCopy.java]: This ResultSet 'rs' and Statement 'stmt' objects are created based on IF condition above, where, Statement might not even be created in 1 condition, hence the finally block is kept as it is.

**Comment by cheenamalhotra** [comment]
Jars for testing : [mssql-jdbc-PR806.zip](https://github.com/Microsoft/mssql-jdbc/files/2421400/mssql-jdbc-PR806.zip)

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDatabaseMetaData.java [hunk 2]
```diff
@@ -265,15 +288,24 @@ private CallableStatement getCallableStatementHandle(CallableHandles request,
             String catalog) throws SQLServerException {
         CallableStatement CS = null;
         HandleAssociation hassoc = handleMap.get(request);
-        if (null == hassoc || null == hassoc.databaseName || !hassoc.databaseName.equals(catalog)) {
-            CS = request.prepare(connection);
-            hassoc = new HandleAssociation(catalog, CS);
-            HandleAssociation previous = handleMap.put(request, hassoc);
-            if (null != previous) {
-                previous.close();
+        try {
```
Comment (peterbae) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDatabaseMetaData.java]: I'm having a hard time understanding the logic change behind HandleAssociation. Can you explain to me why we need another map inside HandleAssociation?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDatabaseMetaData.java [hunk 3]
```diff
@@ -265,15 +288,24 @@ private CallableStatement getCallableStatementHandle(CallableHandles request,
...
+            if (null == hassoc) {
+                CS = request.prepare(connection);
```
Comment (peterbae) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDatabaseMetaData.java]: This CallableStatement used to be closed when this method called previous.close() from line 273. However, we removed this part, and this CS will never be closed. The method the calls this method (getResultSetFromStoredProc) also doesn't clean up this returned CS. Is this CS intentionally left unclosed?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCopy.java [hunk 4]
```diff
@@ -1719,29 +1718,28 @@ private void getDestinationMetadata() throws SQLServerException {
...
-            for (int i = 1; i <= destColumnCount; ++i) {
-                if (rsMoreMetaData.next()) {
-                    if (!connection.getServerSupportsColumnEncryption()) {
-                        destColumnMetadata.put(i, new BulkColumnMetaData(rs.getColumn(i),
-                                rsMoreMetaData.getString("collation_name"), null));
+            try (SQLServerStatement statementMoreMetadata = (SQLServerStatement) connection.createStatement();
+                    SQLServerResultSet rsMoreMetaData = (!connection
+                            .getServerSupportsColumnEncryption() ? statementMoreMetadata
+                                    .executeQueryInternal("select collation_name from sys.columns where "
+                                            + "object_id=OBJECT_ID('" + Util.escapeSingleQuotes(destinationTableName) + "') "
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCopy.java]: This is less readable to me. Maybe you could only "if else" the query string?

Comment (cheenamalhotra) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCopy.java]: Then we can't place it in try with resources block. Also will need another try block, and finally block to manually close it.
I don't see any issues - its a simple ternary operator.

**Pull request merged**